### PR TITLE
Prototype implementation for assessing synthetic data fidelity

### PR DESCRIPTION
**Pull request by Schaechtle**
## What does this do?

Please check out the [README](https://github.com/InferenceQL/lpm.fidelity/blob/schaechtle/prototype-implementation/README.md).

## Why do we want this?

This library is going to help us better understand the quality of synthetic data. As part of an improved QC dashboard, it will help surface best-and-worst fit columns by comparing synthetic to real data.

## How was this tested?

Tests are run automatically if `nix build .` is invoked. A github workflow is going to come in another commit.

### CODE REVIEW

File: flake.nix [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+
+  inputs = {
+    linux-nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.05";
+    darwin-nixpkgs.url = "github:NixOS/nixpkgs/nixos-23.11";
+  };
+
+  outputs = { self, linux-nixpkgs, darwin-nixpkgs, flake-utils, ... }: let
+    pkgs = {
+      x86_64-linux = linux-nixpkgs.legacyPackages.x86_64-linux;
+      aarch64-darwin = darwin-nixpkgs.legacyPackages.aarch64-darwin;
```
Comment (KingMob) [inline flake.nix:12]: I'm on the last model of Intel Mac, so this doesn't work on my machine, fwiw.

```suggestion
      x86_64-linux = linux-nixpkgs.legacyPackages.x86_64-linux;
      aarch64-darwin = darwin-nixpkgs.legacyPackages.aarch64-darwin;
      x86_64-darwin = darwin-nixpkgs.legacyPackages.x86_64-darwin;
```

File: flake.nix [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+            polars
+          ]);
+        });
+    mkShell = pkgs: pkgs.mkShell {
+      buildInputs = [(mkLPMfidelity pkgs)];
+    };
+  in {
+    devShells = {
+      x86_64-linux.default = mkShell pkgs.x86_64-linux;
+      aarch64-darwin.default = mkShell pkgs.aarch64-darwin;
```
Comment (KingMob) [inline flake.nix:44]:
```suggestion
      aarch64-darwin.default = mkShell pkgs.aarch64-darwin;
      x86_64-darwin.default = mkShell pkgs.x86_64-darwin;
```

File: flake.nix [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+    };
+    packages = {
+      x86_64-linux.default = mkLPMfidelity pkgs.x86_64-linux;
+      aarch64-darwin.default = mkLPMfidelity pkgs.aarch64-darwin;
```
Comment (KingMob) [inline flake.nix:49]:
```suggestion
      aarch64-darwin.default = mkLPMfidelity pkgs.aarch64-darwin;
      x86_64-darwin.default = mkLPMfidelity pkgs.x86_64-darwin;
```

File: flake.nix [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+    };
+    apps = {
+      x86_64-linux.default = {
+        type = "app";
+        program = "${mkLPMfidelity pkgs.x86_64-linux}/bin/assess-distance";
+      };
+      aarch64-darwin.default = {
+        type = "app";
+        program = "${mkLPMfidelity pkgs.aarch64-darwin}/bin/assess-distance";
+      };
```
Comment (KingMob) [inline flake.nix:60]:
```suggestion
      };
      x86_64-darwin.default = {
        type = "app";
        program = "${mkLPMfidelity pkgs.x86_64-darwin}/bin/assess-distance";
      };
```

File: flake.nix [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+      x86_64-linux.assess-statistics = {
+        type = "app";
+        program = "${mkLPMfidelity pkgs.x86_64-linux}/bin/assess-statistics";
+      };
+      aarch64-darwin.assess-statistics = {
+        type = "app";
+        program = "${mkLPMfidelity pkgs.aarch64-darwin}/bin/assess-statistics";
+      };
```
Comment (KingMob) [inline flake.nix:72]:
```suggestion
      };
      x86_64-darwin.assess-statistics = {
        type = "app";
        program = "${mkLPMfidelity pkgs.x86_64-darwin}/bin/assess-statistics";
      };      
```

File: src/counting.py [hunk 6]
```diff
@@ -0,0 +1,170 @@
+import polars as pl
+from collections import Counter
+import sys
+import numpy as np
+
+
+def _is_not_none_or_nan(value):
```
Comment (KingMob) [inline src/counting.py]: Minor: I generally find predicate names using "not" to be slightly more awkward to use, especially if you need to negate it somewhere. You end up with "not not_some_pred(x)" which sounds off. Just my $.02.

File: src/two_sample_testing.py [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+from scipy.stats import chi2_contingency
+
+from lpm_fidelity.counting import contingency_table
+
+
+def chi_squared(column_a, column_b):
+    """
+    Compute the Chi^2 contigency table-based two samples test.
+    First, create a sound contingency tables. Then compute the test.
```
Comment (KingMob) [inline src/two_sample_testing.py]: Typo: singular "table", I assume?

File: flake.nix [hunk 8]
```diff
@@ -0,0 +1,65 @@
...
+          ];
+          nativeCheckInputs = with pkgs.python310Packages; [
+            pytest
+            numpy
+            scipy
+            polars
+          ];
+          checkPhase = "pytest tests/ -vvv";
+          propagatedBuildInputs = (with pkgs.python310Packages; [
+            numpy
```
Comment (KingMob) [inline flake.nix:30]:
```suggestion
            pytest
            numpy
```

The directions in the README, to run the tests with `nix develop -c  pytest tests/ -vvv`, fails for me unless pytest is added to `propagatedBuildInputs`

**Review by KingMob — COMMENTED** [review]
Generally LGTM, tho I'm not well-versed in Python. 

I have some minor suggestions for supporting Intel macs, and I think the nix pytest command is broken without specifying it in the flake.nix. (Or having it installed globally.)

File: flake.nix (hunk 8 above)
Comment (Schaechtle) [reply flake.nix:30]: ah right. I think I want `pytest in the output of the devshell, then.

Comment (Schaechtle) [reply flake.nix:30]: done.

**Pull request merged**